File: solutions/python/list-ops/1/list_ops.py
def append(list1, list2):
    for n in list2:
        list1.append(n)
    return list1

def concat(lists):
    res = []
    for n in lists:
        for nn in n:
            res.append(nn)
    return res
        

def filter(function, list):
    res = []
    for x in list:
        if function(x):
            res.append(x)
    return res


def length(list):
    x = 0
    for _ in list:
        x += 1
    return x


def map(function, list):
    res = []
    for n in list:
        res.append(function(n))
    return res

def foldl(function, list, initial):
    for n in list:
        initial = function(initial,n)
    return initial


def foldr(function, list, initial):
    for n in reverse(list):
        initial = function(initial,n)
    return initial


def reverse(list):
    return list[::-1]

File: solutions/javascript/list-ops/1/list-ops.js
//
// This is only a SKELETON file for the 'List Ops' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class List {
  constructor(values = []) {
    this.values = values || [];
  }

  append(list) {
    if (list.length() < 0) return new List([...this.values]);
    return new List([...this.values, ...list.values]);
  }

  concat(arrLists) {
    let concatList = new List(this.values);
    if (arrLists.length() > 0) {
      for (let list of arrLists.values) {
        concatList = concatList.append(list);
      }
    }

    return concatList;
  }

  filter(fn) {
    let filterList = new List();
    for (let elemento of this.values)
      if (fn(elemento)) filterList = filterList.append(new List([elemento]));
    return filterList;
  }

  map(fn) {
    let mapList = new List();
    for (let elemento of this.values)
      mapList = mapList.append(new List([fn(elemento)]));
    return mapList;
  }

  length() {
    let len = 0;
    for (let x of this.values) len++;
    return len;
  }

  foldl(fn, ac) {
    for (let elemento of this.values) ac = fn(ac, elemento);
    return ac;
  }

  foldr(fn, ac) {
    for (let i = this.length() - 1; i >= 0; i--) ac = fn(ac, this.values[i]);
    return ac;
  }

  reverse() {
    let reverseList = new List();
    for (let i = this.length() - 1; i >= 0; i--) {
      reverseList = reverseList.append(new List([this.values[i]]));
    }
    return reverseList;
  }
}
